crypto/core_fetch.c: before contructing a method, check if it already exists

levitte opened the pull request:
Fixes #9561

levitte committed:
crypto/core_fetch.c: before contructing a method, check if it already exists

Fixes #9561

levitte committed:
fixup! crypto/core_fetch.c: before contructing a method, check if it already exists

bernd-edlinger commented on crypto/core_fetch.c:
```diff
@@ -74,6 +75,7 @@ static void ossl_method_construct_this(OSSL_PROVIDER *provider,
                    algo->algorithm_name, algo->property_definition,
```
I wonder why you are not looking at the return value of put ?

levitte replied:
Read the comment at (new) lines 53-62

bernd-edlinger replied:
Ah, okay, maybe explicitly casting to void would make sense then.

levitte replied:
Done

bernd-edlinger commented on crypto/core_fetch.c:
```diff
@@ -31,6 +31,20 @@ static void ossl_method_construct_this(OSSL_PROVIDER *provider,
+     * to copy a property definition and paste that wherever a property
+     * is being queried.
+     */
+    if ((method = data->mcm->get(data->libctx, NULL, data->operation_id,
```
I kind of wonder, if you should also try to `data->mcm->get(data->libctx, data->store, ...)` here,
especially when the put to NULL-store is optional?

levitte replied:
Why?  `data->store` is a temporary store, so that would really only be useful if the provider has more than one implementation with the same \< name, properties \> pair.  This *might* of course be an issue, but in that case, I would prefer to fix that in a separate PR.

bernd-edlinger replied:
I really don't know this code good enough, I am just wondering.
Do you try to get the method from a global store even when this
the constructor is not going to add it to the global store `!((data->force_store || !no_store)`

levitte replied:
You do have a point that it's pretty pointless to look for it in the global store in that case.  Ok, will fix

bernd-edlinger replied:
I kind of wonder if this sequence needs a lock, especially when duplicates
in the store create a memory leak ?

bernd-edlinger replied:
why was there a memory leak in the first place, I mean isn't the put method
"expected" to do the right thing ?

bernd-edlinger commented on crypto/core_fetch.c:
```diff
@@ -31,6 +31,20 @@ static void ossl_method_construct_this(OSSL_PROVIDER *provider,
+     * property query, which stresses that they MUST have compatible
+     * syntax, or else...  (in other words, it must always be possible
+     * to copy a property definition and paste that wherever a property
+     * is being queried.
```
Sorry, but i don't understand the second part of this comment at all.
That may be an indication that I don't understand this code very well.

levitte replied:
You'll have to look at the manual for properties: https://github.com/openssl/openssl/blob/master/doc/man7/property.pod
The SYNTAX section is of interest here.  You have see that there's a `Definition` and a `Query` syntax, and they share certain properties, such as having `=` for assignment as well as for equality test.  This means that the *property definition* `foo=yes, bar=no` and the *property query* `foo=yes, bar=no` look exactly the same.  Now, the `put()` method takes a *property definition*, while the `get()` method takes a *property query*, but since the syntax is compatible, we can use the *property definition* as a *property query* to find the method that matches the exact \< name, properties \> pair that the provider gave us here.

If the *property query* syntax would change to use `==` for equality checks, this trick will no longer work, and we will need some other mechanism to "get" a method with an exact name and property definition.

@paulidale might have a word or two to say here...

levitte replied:
I rewrote the comment, in the hope that it's easier to understand

bernd-edlinger replied:
Ah, interesting, that is...

paulidale replied:
I'm not sure what to say.  It works and was designed this way, I don't see why it should ever not work (but I've been wrong before).

It would be possible to add a call to convert a definition to a query, it would be difficult to guarantee that the same algorithm was returned when the query was used.

levitte committed:
fixup! crypto/core_fetch.c: before contructing a method, check if it already exists

bernd-edlinger commented:
BTW: s/contructing/constructing/ in the commit message.

bernd-edlinger commented:
could it be that there is a memory leak in the put method?
https://github.com/openssl/openssl/blob/1a2a3a420609c2d540b544e5ae829a3cb468a6e3/crypto/evp/evp_fetch.c#L135-L139

bernd-edlinger commented on crypto/core_fetch.c:
```diff
@@ -30,7 +30,31 @@ static void ossl_method_construct_this(OSSL_PROVIDER *provider,
+     * This is possible because property definitions have a syntax that
+     * can also be used for a property query with equality tests.  We do
+     * this to look up already existing methods matching the implementation
+     * that we just for from the provider (via |algo|),
```
much better this way, lust the last few words need to be unscrambled
`that we just for from the provider` and should this sentence end with a period ?

levitte replied:
Typo.  Fixed.

bernd-edlinger commented on crypto/core_fetch.c:
```diff
@@ -30,7 +30,31 @@ static void ossl_method_construct_this(OSSL_PROVIDER *provider,
+     * supposed to store the method in the global store.
+     *
+     * Note that while data->mcm->put() takes a property *definition*,
+     * while data->mcm->get() takes a property *query*, but we're using
```
I think the second while is superfluous, but @paulidale will know better.

levitte replied:
Yes, it is superfluous.

levitte committed:
fixup! crypto/core_fetch.c: before contructing a method, check if it already exists

levitte committed:
crypto/evp/evp_fetch.c: fix potential memory leak

If the refcount was successfully incremented, but adding the method to
the store failed, we end up having too high a refcount, so we need to
lower it again.

bernd-edlinger commented on crypto/core_fetch.c:
```diff
@@ -30,7 +30,31 @@ static void ossl_method_construct_this(OSSL_PROVIDER *provider,
+                                    algo->algorithm_name,
+                                    algo->property_definition,
+                                    data->mcm_data)) != NULL)
+        goto end;
```
So if the method exists in the global store, we are not going to add it to the local store, right?
But why are we adding it to the local store at all?

levitte replied:
We're adding it to a local store so `ossl_method_construct()` can find it after we've added all methods that the providers had to offer (see (new) line 117).  This fullfils two objectives: 1) only do a second lookup on stuff that is newly added, 2) do a second lookup on methods that aren't supposed to remain in any store.

bernd-edlinger replied:
but when the method happens to exist in the global store it will not be added
to the local store which mitght be a different one ?

levitte replied:
If it does, it was already picked up by `ossl_method_construct`.

bernd-edlinger replied:
So why don't you add the method from the global store to the local store, so ossl_method_construct can
pick it up?

bernd-edlinger replied:
or does this execute under some kind of global mutex?

levitte replied:
> So why don't you add the method from the global store to the local store, so `ossl_method_construct` can pick it up?

Like I said, `ossl_method_construct` has already picked it up.  It's the first thing it tries to do, and it only calls `ossl_method_construct_this` via `ossl_algorithm_do_all` when the implementation isn't in the global store at that time.  It's pointless to copy methods from the global store to the temporary store after that.

However, it's true that this isn't exactly thread safe.  In other words, we currently expact `FOO_fetch()` to be called in a single thread...  I'm unsure if we're explicit about that, though...  I think that's worth a separate issue if it turns out that we should be more protective.

bernd-edlinger commented on crypto/core_fetch.c:
```diff
@@ -47,20 +71,18 @@ static void ossl_method_construct_this(OSSL_PROVIDER *provider,
-    data->mcm->put(data->libctx, data->store, method, data->operation_id,
-                   algo->algorithm_name, algo->property_definition,
-                   data->mcm_data);
+    (void)data->mcm->put(data->libctx, data->store, method, data->operation_id,
```
I mean, should this be `} else {` ?

levitte replied:
No.  That would mean that `ossl_method_construct` will never find newly added methods.

bernd-edlinger replied:
but this will be broken by this PR, since you skip that function when the
method is already in the global store, right?

bernd-edlinger replied:
So unless ossl_method_construct executes under some kind of write_lock,
we can have a race condition making a second thread executing
the same ossl_method_construct fail.

bernd-edlinger commented on crypto/evp/evp_fetch.c:
```diff
@@ -132,10 +132,13 @@ static int put_method_in_store(OPENSSL_CTX *libctx, void *store,
-                                 methdata->destruct_method))
-        return 1;
+    if (methdata->refcnt_up_method(method)) {
+        if (ossl_method_store_add(store, methid, propdef, method,
```
I kind of wonder, if this duplicate detection should be done on the ossl_method_store_add
where we have a lock for that?

levitte replied:
That's a refactoring question that I'd rather keep out of this PR.

bernd-edlinger replied:
I start to think that the duplicate detection should be in `ossl_method_store_add`
and not in `ossl_method_construct_this`.
Why should the store add the same object twice?  Is there a use case for it?

levitte replied:
The "local" store is *temporary*

levitte commented:
Some of @bernd-edlinger's remarks made me think another round...  he's right that it should be possible for `ossl_method_store_add` to detect if a method with a specific set of properties was already added...  or something similar.  So this PR will see some rework

levitte commented:
Closing in favor of #9650
